src/zkevm_specs/pi_circuit.py

```diff
@@ -91,62 +137,43 @@ class PublicInputs:
 def check_row(
     row: Row,
     row_next: Row,
-    row_offset_block_table_value_hi: Row,
-    row_offset_tx_table_tx_id: Row,
-    row_offset_tx_table_index: Row,
-    row_offset_tx_table_value_lo: Row,
-    row_offset_tx_table_value_hi: Row,
-    table: Set[TxCallDataGasCostAccRow],
+    calldata_gas_cost_table: Set[TxCallDataGasCostAccRow],
     fixed_u16_table: Set[FixedU16Row],
+    keccak_table: KeccakTable,
+    circuit_len: FQ,
 ):
-    q_not_end = row.q_not_end
-    q_end = row.q_end
-    row_offset_block_table_value_lo = row
-
-    # 0.0 rpi_rlc_acc[0] == RLC(raw_public_inputs, rand_rpi)
-    assert q_not_end * row.rpi_rlc_acc == q_not_end * (
-        row_next.rpi_rlc_acc * row.rand_rpi + row.raw_public_inputs
-    )
-
-    assert q_end * row.rpi_rlc_acc == q_end * row.raw_public_inputs
+    q_bytes_last = row.q_bytes_last
+    q_rpi_byte_enable = row.q_rpi_byte_enable
 
-    # 0.1 rand_rpi[i] == rand_rpi[j]
-    assert q_not_end * row.rand_rpi == q_not_end * row_next.rand_rpi
+    # gate 1 and gate 2 are compensation branch
+    # 1: rpi_bytes_keccakrlc[last] = rpi_bytes[last]
+    assert q_rpi_byte_enable * q_bytes_last * (row.rpi_bytes_keccakrlc - row.rpi_bytes) == FQ(0)
 
-    # 0.2 Block table -> value column match with raw_public_inputs at expected offset
-    assert (
-        row.q_block_table * row.block_table.value.lo.expr()
-        == row.q_block_table * row_offset_block_table_value_lo.raw_public_inputs
-    )
+    # 2: rpi_bytes_keccakrlc[i] = keccak_rand * rpi_bytes_keccakrlc[i+1] + rpi_bytes[i]"
     assert (
-        row.q_block_table * row.block_table.value.hi.expr()
-        == row.q_block_table * row_offset_block_table_value_hi.raw_public_inputs
+        q_rpi_byte_enable
+        * (FQ(1) - q_bytes_last)
+        * (row.rpi_bytes_keccakrlc - row_next.rpi_bytes_keccakrlc * keccak_rand - row.rpi_bytes)
```

Comment:
```suggestion
        q_rpi_byte_enable
        * (FQ(1) - q_bytes_last)
        * (row.rpi_bytes_keccakrlc - (row_next.rpi_bytes_keccakrlc * keccak_rand + row.rpi_bytes))
        == FQ(0)
```
A suggestion to make it easier to follow the equality constraint: Instead of `a - b - c` do `a - (b + c)`
Also this assert contains an FQ and not boolean so it needs `== FQ(0)` at the end.

Reply:
The original part parantheses is added, but I noticed there are 2 formula and another one is missing :), Done in latest commit https://github.com/privacy-scaling-explorations/zkevm-specs/pull/417/commits/57af8f62081ddaf4fe7964f2f2496b933d814b93